[spec] add environment variables as context propagators specification (#4454)
open-telemetry/opentelemetry-specification: specification/context/env-carriers.md

Author: Great question, I'm not sure of the answer.

Reviewer: We don't have a good place for the *standard* on format here.  E.g. api-propagators defines mechanism, not format.

I think this is fine here and a next PR should find how to "advertise" this capability in other parts of the spec, but you could also add that in this PR.

Author: I took a stab at putting it in some minimally invasive, but readily available spots. Going to close this thread. Please feel free to give it a re-review and re-open this one, or open a new one if there's additional suggestions/requests!

applicable. This specification extends the [API Propagators](../context/api-propagators.md)
to define how the
[TextMapPropagator](../context/api-propagators.md#textmap-propagator) can be
used with environment variables.

Common systems where context propagation via environment variables is useful
include:

- Batch processing systems
- CI/CD environments
- Command-line tools

## Propagator Mechanisms

Propagating context via environment variables involves reading and writing to
environment variables. A `TextMapPropagator` SHOULD be used alongside its
normal `Get`, `Set`, `Extract`, and `Inject` functionality as described in the [API
Propagators](../context/api-propagators.md) specification.

### Environment Variable Names

It is RECOMMENDED to use the [W3C Trace
Context](https://www.w3.org/TR/trace-context/) and [W3C
Baggage](https://www.w3.org/TR/baggage/) specifications mapped to environment
variable names for consistent context propagation.

When using the W3C Trace Context and Baggage propagators with environment
variables, the following translated standard environment variable names SHOULD
be used:

| Context Information | Environment Variable | W3C Header Equivalent |

> |---------------------|----------------------|------------------------|
> | Trace Context       | `TRACEPARENT`        | `traceparent`          |
> | Trace State         | `TRACESTATE`         | `tracestate`           |
> | Baggage             | `BAGGAGE`            | `baggage`              |

Reviewer 2: ```suggestion
|---------------------|----------------------|-----------------------|
| Trace Context       | `TRACEPARENT`        | `traceparent`         |
| Trace State         | `TRACESTATE`         | `tracestate`          |
| Baggage             | `BAGGAGE`            | `baggage`             |
```

Implementations MAY support additional propagation formats and SHOULD provide
configuration options to override the default environment variable.

### Format Restrictions

#### Name Restrictions

Environment variable names used for context propagation:

- SHOULD use uppercase letters, digits, and underscores for maximum
  cross-platform compatibility
- MUST NOT include characters forbidden in environment variables per
  platform-specific restrictions
- SHOULD follow naming conventions that align with the propagation format
  specification they're implementing (e.g., `TRACEPARENT` for W3C trace context)

#### Value Restrictions

Environment variable values used for context propagation:

- MUST only use characters that are valid in HTTP header fields per [RFC
  9110](https://tools.ietf.org/html/rfc9110)
- MUST follow the format requirements of the specific propagation protocol
  (e.g., W3C Trace Context specification for `TRACEPARENT` values)
- SHOULD NOT contain sensitive information

#### Size Limitations

Implementations SHOULD follow platform-specific environment variable size

> limitations:

Reviewer 2: Consider borrowing some idea/wording from https://www.w3.org/TR/trace-context/#tracestate-limits.

Author: Thanks for the reference. i've updated so will resolve this comment. Feel free to open another one based on the changes!

- Windows: Maximum 32,767 characters for name=value pairs according to
  [Microsoft Documentation](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-setenvironmentvariable)
- UNIX: System-dependent limits exist and are typically lower than Windows.

Implementations SHOULD handle truncation gracefully when size limits are
exceeded.

### Operational Guidance

#### Environment Variable Immutability

Once set for a process, environment variables SHOULD be treated as immutable
within that process:

- Applications SHOULD read context-related environment variables during
  initialization.

> - Applications SHOULD NOT modify environment variables on initialization of
>   environment in which the parent process exists.

Reviewer 3: ```suggestion
- Applications SHOULD NOT modify context-related environment variables of the
  environment in which the parent process exists.
```

- If modification is required, applications SHOULD create a new environment

>   variables dictionary with the updated values for the child process.

Reviewer 2: What does "dictionary" mean here?

Author: edit: in hind-site, I think this doesn't make sense to be here and is confusing as written. Instead the direction and intent is covered correctly in the process spawning section immediately below.

Author: I've removed the line, and update the process spawning section to add a little more clarity on the meaning.

#### Process Spawning

When spawning child processes:

- Parent processes SHOULD inject context into environment variables before
  spawning child processes.
- Child processes SHOULD extract context from environment variables at startup.
- When spawning multiple child processes with different contexts or baggage,
  each child SHOULD receive its own copy of the environment variables with
  appropriate information.

#### Security

Environment variables are generally accessible to all code running within a
process and with the correct permissions, can be accessed from other processes.

- Implementations SHOULD NOT store sensitive information in environment
  variables.
- Applications running in multi-tenant environments SHOULD be aware that
  environment variables may be visible to other processes or users with
  appropriate permissions.

#### Case Sensitivity

Environment variable names are case-sensitive on UNIX and case-insensitive on
Windows.

> - For maximum compatibility, implementations SHOULD:

Reviewer 2: Why "SHOULD" instead of "MUST"?

Author: Great call out. I tried to stay away from MUST as much as possible, but I think (if I'm correctly inferring the reasoning behind your question) that MUST would be better suited here. Will make the update.

Author: Updated

Reviewer 2: Here is the context, in the spec we follow the RFC wording https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification#notation-conventions-and-compliance (no worry about "being polite" in the spec wording when it comes to "MUST" vs "SHOULD").

    + [Security](#security)
    + [Case Sensitivity](#case-sensitivity)

<!-- tocstop -->

</details>

## Overview

Environment variables provide a mechanism to propagate context and baggage
information across process boundaries when network protocols are not
applicable. This specification extends the [API Propagators](../context/api-propagators.md)
to define how the
[TextMapPropagator](../context/api-propagators.md#textmap-propagator) can be
used with environment variables.

Common systems where context propagation via environment variables is useful
include:

- Batch processing systems
- CI/CD environments
- Command-line tools

## Propagator Mechanisms

Propagating context via environment variables involves reading and writing to
environment variables. A `TextMapPropagator` SHOULD be used alongside its
normal `Get`, `Set`, `Extract`, and `Inject` functionality as described in the [API
Propagators](../context/api-propagators.md) specification.

### Environment Variable Names

It is RECOMMENDED to use the [W3C Trace
Context](https://www.w3.org/TR/trace-context/) and [W3C
Baggage](https://www.w3.org/TR/baggage/) specifications mapped to environment

> variables names for consistent context propagation.

Reviewer 3: Should this be without `s`?
```suggestion
variable names for consistent context propagation.
```

  specification they're implementing (e.g., `TRACEPARENT` for W3C trace context)

#### Value Restrictions

Environment variable values used for context propagation:

- MUST only use characters that are valid in HTTP header fields per [RFC
  9110](https://tools.ietf.org/html/rfc9110)
- MUST follow the format requirements of the specific propagation protocol
  (e.g., W3C Trace Context specification for `TRACEPARENT` values)
- SHOULD NOT contain sensitive information

#### Size Limitations

Implementations SHOULD follow platform-specific environment variable size
limitations:

- Windows: Maximum 32,767 characters for name=value pairs according to
  [Microsoft Documentation](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-setenvironmentvariable)
- UNIX: System-dependent limits exist and are typically lower than Windows.

Implementations SHOULD handle truncation gracefully when size limits are
exceeded.

### Operational Guidance

#### Environment Variable Immutability

Once set for a process, environment variables SHOULD be treated as immutable
within that process:

- Environment variables SHOULD only be modified before the creation of one or

>   more child processes occur.

Reviewer 3: Does this include modifying the environment variable of the parent process?
Doesn't the point below sufficiently cover modification of env vars?
> - If modification is required, applications SHOULD create a new environment
  variables dictionary with the updated values for the child process.

Author: I think the wording needs adjusted. It shouldn't be modifying the original, in as much as it's copying the original, making modifications, and then spawning it within the creation of the child process.

- Applications SHOULD NOT modify context-related environment variables during
  execution within the environment in which they were created, as this can lead

>   to inconsistent trace contexts.

Reviewer 3: Isn't this very difficult to enforce?
Can we mitigate against changes of the env vars by saying that any implementation of `EnvVarPropagator` should only read the context-related environment variables once during initialization?

Author: I think the wording can be adjusted to be more clear, yes.

# Environment Variables as Context Propagation Carriers

**Status**: [Experimental](../document-status.md)

<details>
<summary>Table of Contents</summary>

<!-- toc -->

- [Overview](#overview)
- [Propagator Mechanisms](#propagator-mechanisms)
  * [Environment Variable Names](#environment-variable-names)
  * [Format Restrictions](#format-restrictions)
    + [Name Restrictions](#name-restrictions)
    + [Value Restrictions](#value-restrictions)
    + [Size Limitations](#size-limitations)
  * [Operational Guidance](#operational-guidance)
    + [Environment Variable Immutability](#environment-variable-immutability)
    + [Process Spawning](#process-spawning)
    + [Security](#security)

>     + [Case Sensitivity](#case-sensitivity)

Reviewer 2: I think this PR has covered how things can be propagated from a parent process to and child process, and what's the actual protocol.
From the API/SDK spec point of view, this PR hasn't covered what the language implementation SIGs should do - for example, what type and API should be provided? If you look at https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md, it has defined `TextMapPropagator`, the `Inject` and `Extract` operations.

Author: Thanks for the input! The intent of the statement below was to signify that SDK's would use the already existing TextMapPropagator, and it's operations but in the context of reading from the environment instead of through headers in this scenario. 

> This specification extends the [API Propagators](../context/api-propagators.md)
> to define how the
> [TextMapPropagator](../context/api-propagators.md#textmap-propagator) can be
> used with environment variables.

Both the [swift](https://github.com/open-telemetry/opentelemetry-swift/blob/main/Sources/OpenTelemetrySdk/Trace/Propagation/EnvironmentContextPropagator.swift#L13) and [python](https://github.com/Div95/opentelemetry-python/blob/feature/env_propagator/propagator/opentelemetry-propagator-env/src/opentelemetry/propagators/env/propagator.py#L33) extend the TextMapPropagator. 

Is saying the current API propagator definitions for TextMap propagation are the standard enough? Or should I add something like:

> An `EnvPropagator` extends the `TextMapPropagator` and its operations, but in the context of an environment? 

Open to suggestions for sure. I was hoping to avoid redeclaring the operations since they're the same and should abide by the API Propagators spec.

Reviewer 2: > Or should I add something like:
> 
> > An `EnvPropagator` extends the `TextMapPropagator` and its operations, but in the context of an environment?
> 
> Open to suggestions for sure. I was hoping to avoid redeclaring the operations since they're the same and should abide by the API Propagators spec.

Your proposal sounds good to me with one caveat - I would vote for something called "EnvVarPropagator" or "EnvironmentVariablePropagator" or "EnvironmentVariablesPropagator".

Also, it might be helpful to give some concrete examples for the language implementation owners to provide language idiomatic experience (for reference, here is a good example https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#histogram).

Reviewer 4: For me this is the only remaining question: whether or not to provide decorating propagators that automatically handle the env-var part (while delegating the actual serialization to the w3c propagators, as shown in the prototypes).

I wouldn't be opposed to deferring adding those EnvVarPropagators, but no strong opinion. But at least we should do as @reyang says, and add a few more concrete examples.

Author: I like it, makes sense & agreed! If no objections, let's merge this pull request and I'll open an issue to track adding EnvVar decorators.

Reviewer 2: Merged.
